File: src/main/java/january_2025/DecodeWays.java
package january_2025;

import java.util.HashMap;
import java.util.Map;

public class DecodeWays {

    // O(n) time | O(n) space
    public int numDecodings(String s) {
        Map<Integer, Integer> dp = new HashMap<>();
        dp.put(s.length(), 1);

        return dfs(0, s, dp);
    }

    private int dfs(int i, String s, Map<Integer, Integer> dp) {
        if (dp.containsKey(i)) {
            return dp.get(i);
        }
        if (s.charAt(i) == '0') {
            return 0;
        }
        int res = dfs(i + 1, s, dp);
        if (i + 1 < s.length() && (s.charAt(i) == '1' || s.charAt(i) == '2' && s.charAt(i + 1) < '7')) {
            res += dfs(i + 2, s, dp);
        }
        dp.put(i, res);
        return res;
    }

}

File: python/decode_ways.py
class Solution(object):
    # O(n) time | O(n) space
    def numDecodings(self, s):
        """
        :type s: str
        :rtype: int
        """
        dp = {}
        dp[len(s)] = 1

        def dfs(i):
            if i == len(s):
                return 1
            if s[i] == "0":
                return 0
            if i in dp:
                return dp[i]
            
            res = dfs(i + 1)
            if i + 1 < len(s) and (s[i] == "1" or (s[i] == "2" and s[i + 1] in "0123456")):
                res += dfs(i + 2)
            dp[i] = res
            return res

        return dfs(0)
    

solution = Solution()
print(solution.numDecodings("226"))


